src/main/java/ca/uhn/fhir/jpa/starter/interceptor/LoggerInterceptor.java

```diff
@@ -15,32 +12,54 @@
 import ca.uhn.fhir.rest.api.server.RequestDetails;
 import ca.uhn.fhir.rest.api.server.ResponseDetails;
 import ca.uhn.fhir.rest.server.servlet.ServletRequestDetails;
+import jakarta.servlet.http.HttpServletRequest;
+import jakarta.servlet.http.HttpServletResponse;
 
 @Interceptor
-public class CorrelationIdInterceptor {
+public class LoggerInterceptor {
   private static final String X_CORRELATION_ID = "X-Correlation-Id";
+	private static final String CORRELATION_ID = "correlationId";
+	private static final String X_LAUNCH_ID = "X-Launch-Id";
+  private static final String X_APP_NAME = "X-App-Name";
+  private static final String LAUNCH_ID = "launchId";
+  private static final String APP = "app";
+
 
   @Hook(Pointcut.SERVER_INCOMING_REQUEST_PRE_PROCESSED)
   public boolean incomingRequestPreProcessed(HttpServletRequest request,
       HttpServletResponse response) {
-    String corelationId = ObjectUtils.isNotEmpty(request.getHeader(X_CORRELATION_ID))
+	 String correlationId = ObjectUtils.isNotEmpty(request.getHeader(X_CORRELATION_ID))
         ? request.getHeader(X_CORRELATION_ID)
         : UUID.randomUUID().toString();
-    MDC.put(X_CORRELATION_ID, corelationId);
-    return true;
+	MDC.put(CORRELATION_ID, correlationId);
+		putIfPresent(request, X_LAUNCH_ID, LAUNCH_ID);
+		putIfPresent(request, X_APP_NAME, APP);
+
+		return true;
   }
 
+	private void putIfPresent(HttpServletRequest request, String headerName, String mdcKey) {
+		String headerValue = request.getHeader(headerName);
+		if (ObjectUtils.isNotEmpty(headerValue)) {
+			MDC.put(mdcKey, headerValue);
+		}
+	}
+
   @Hook(Pointcut.SERVER_OUTGOING_RESPONSE)
   public boolean outgoingResponse(RequestDetails requestDetails,
       ServletRequestDetails servletRequestDetails, IBaseResource resource, ResponseDetails responseDetails,
       HttpServletRequest request, HttpServletResponse response) {
-    response.addHeader(X_CORRELATION_ID, MDC.get(X_CORRELATION_ID));
-    return true;
+    response.addHeader(CORRELATION_ID, MDC.get(CORRELATION_ID));
+	  String launchId = MDC.get(LAUNCH_ID);
+	  if (launchId != null) {
+		  response.addHeader(LAUNCH_ID, launchId);
+	  }
+	  return true;
   }
 
   @Hook(Pointcut.SERVER_PROCESSING_COMPLETED)
   public void processingCompleted(RequestDetails requestDetails,
       ServletRequestDetails servletRequestDetails) {
-    MDC.remove(X_CORRELATION_ID);
+	  MDC.clear();
```

Comment:
Why `clear` instead of just removing the `correlationId`?

Reply:
We have added `correlationId`, `launchId`, `app` in MDC now, so I have used `MDC.clear();` to clear all of them after process is completed.

Reply:
Isn't is possible that there will be other items in the `MDC` that we didn't add and shouldn't be clearing?

Reply:
Ok, I'll remove `MDC.clear();`

Reply:
Are you sure the hapi-fhir code is not using MDC? Are you sure they will not in the future?

I would think the best practice is to clean up things you created - not clear.

Reply:
Ok, I'll clean up the things I created.